telemetry/data_slow_query.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telemetry
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	pingcapErrors "github.com/pingcap/errors"
+	"github.com/pingcap/tidb/domain/infosync"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/prometheus/client_golang/api"
+	promv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	pmodel "github.com/prometheus/common/model"
+	"go.uber.org/zap"
+)
+
+type slowQueryStats struct {
+	// Slow Query statistic buckets
+	SQBInfo *SlowQueryBucket `json:"slowQueryBucket"`
+}
+
+// SlowQueryBucket records the statistic information of slow query buckets
+// Buckets:   prometheus.ExponentialBuckets(0.001, 2, 28), // 1ms ~ 1.5days  // defined in metrics/server.go
+type SlowQueryBucket map[string]int
+
+func (bucketMap SlowQueryBucket) String() string {
+	if bucketMap == nil {
+		return "nil"
+	}
+	var retStr string = "{"
+	for k, v := range bucketMap {
+		retStr += k + ":" + strconv.Itoa(v) + ","
+	}
+	retStr = retStr[:len(retStr)-1]
+	return retStr
+}
+
+const slowQueryBucketNum = 29 //prometheus.ExponentialBuckets(0.001, 2, 28), and 1 more +Inf
+
+var (
+	// LastSQBInfo records last statistic information of slow query buckets
+	LastSQBInfo SlowQueryBucket
+	// CurrentSQBInfo records current statitic information of slow query buckets
+	CurrentSQBInfo SlowQueryBucket
+	slowQueryLock  = sync.RWMutex{}
+)
+
+func getSlowQueryStats(ctx sessionctx.Context) (*slowQueryStats, error) {
+	slowQueryLock.Lock()
+	slowQueryBucket, err := getSlowQueryBucket(ctx)
+	slowQueryLock.Unlock()
```

Comment:
I suggest to apply the lock only at where slowQueryBucket variable is modified. Otherwise the lock internal is too large, e.g. there is even a network request inside, will hold lock for long time.

Reply:
Accepted.
Done.